feat: add documentation around deployment to kubernetes (#397)
amplication/docs: docs/how-to/deploy-kubernetes.md

slug: /deploy/kubernetes
---

In the dynamic realm of modern software development, deploying applications on Kubernetes has become the defacto standard for reliability, flexibility and faster iteration. This guide distills the process of deploying an Amplication generated application on a Kubernetes cluster.

## Pre-requisites

Although default amplication code generation provides a few of the building blocks required to be able to deploy to kubernetes - e.g., Dockerfile. We will supplement the code base with the other pre-requisites by installing two plugins on the applicable service. These plugins are `Github Actions` and `Helm Chart`. Using these plugins will result in having the artifacts required to deploy the application to Kubernetes - a container image & helm chart.

### Step 1: Configuration of the Github Actions plugin

When in the overview of the to be deployed service, navigate to the plugins part of the service configuration. Here the `Github Actions` plugin can be found in between various other plugins. Press the `Install` button, this should introduce pending changes and the option to press a `Settings` button, press this to be able to configure the plugin further.

> The setting `registry` determines the workflow base that is used. When it is left empty, the default workflow is used - this workflow only includes steps for building and testing the service.

Reviewer: Minor change to the phrase:

```suggestion
The setting `registry` determines the workflow base that is used. When it is left empty, the default workflow, that only includes steps for building and testing the service, is used.
```

When in the overview of the to be deployed service, navigate to the plugins part of the service configuration. Here the `Helm Chart` plugin can be found. Press the `Install` button, this should introduce pending changes and the option to press a `Settings` button, press this to be able to configure the plugin further. As with the `Github Actions` plugin some changes are made to the default generated json settings.

The `root_level` setting determines whether the directory for the helm charts is placed at the root of the repository or in the base directory of the service. The `directory_name` setting determines what the sub-directory for the helm chart in the root level or service base directory is called.

```json title="helm-chart-plugin/settings" showLineNumbers
{
  "root_level": true,
  "directory_name": "helm",
  "server": {
    "chart_version": "0.0.1",
    "application_version": "0.0.1",
    "repository": "ghcr.io/amplication/example-kubernetes-deployment/logists",
    "tag": "latest",
    "host": "server.example.com",
    "port": "3000"
  },
  "admin_ui": {
    "enabled": false
  }
}
```

:::note

> Everything that is in the environments variable file for the service is moved to the `configmap` part of the helm chart, it would be advised to move secret related configuration to the secrets object and preferably not have the secrets stored in the generated code at all (as this is implementation specific the decision was made to add everything to the `configmap`).

Reviewer: ```suggestion
Everything that is in the environments variable file for the service is moved to the `configmap` part of the helm chart. As this might contain also secrets, we advise to move secret related configuration to the secrets object and preferably not have the secrets stored in the generated code at all (as this is implementation specific the decision was made to add everything to the `configmap`).
```
